PolyEngine/Core/Src/Collections/BitMask.cpp

```diff
@@ -0,0 +1,382 @@
+#include "CorePCH.hpp"
+
+#include <climits>
+
+constexpr u64 ZERO = 0UL;
+
+using namespace Poly;
+
+constexpr BitMask::DataType TYPE_BIT = CHAR_BIT * sizeof(BitMask::DataType);
+
+BitMask::BitMask(size_t size)
+	: BitsNumber(size)
+{
+	size_t arraySize = 0;
+
+	if (size%TYPE_BIT)
+		arraySize = size / TYPE_BIT + 1;
+	else
+		arraySize = size / TYPE_BIT;
+
+	for (size_t i = 0; i < arraySize; i++)
+		BitList.PushBack(ZERO);
+}
+
+bool BitMask::Reset()
+{
+	BitList.Clear();
+	BitsNumber = 0;
+	return true;
+}
+
+bool BitMask::Toggle(size_t index)
+{
+	HEAVY_ASSERTE(index <= BitsNumber, "Out of bounds");
+	BitList[index / TYPE_BIT] ^= 1UL << index%TYPE_BIT;
+	return true;
+}
+
+bool BitMask::operator[](size_t index) const
+{
+	DataType tempChar = BitList[index / TYPE_BIT];
+	tempChar = (tempChar >> index%TYPE_BIT) & 1UL;
+	return tempChar != 0;
+}
+
+bool BitMask::Set(size_t index, bool state)
+{
+	HEAVY_ASSERTE(index<=BitsNumber, "Out of bounds");
+
+	size_t bitListIndex = index / TYPE_BIT;
+	size_t bitPosition = index % TYPE_BIT;
+
+	if (state)
+		BitList[bitListIndex] |= (1UL << bitPosition);
+	else
+		BitList[bitListIndex] &= ~(1UL << bitPosition);
+	return true;
+}
+
+BitMask BitMask::operator|(const BitMask& rhs) const
+{
+	//Equal Dynarray sizes
+	if (BitList.GetSize() == rhs.BitList.GetSize())
+	{
+		size_t tempBitsNumber = 0;
+		if (BitsNumber > rhs.BitsNumber)
+			tempBitsNumber = BitsNumber;
+		else
+			tempBitsNumber = rhs.BitsNumber;
+
+		BitMask temp(tempBitsNumber);
+		for (size_t i = 0; i < BitList.GetSize(); i++)
+			temp.BitList[i] = BitList[i] | rhs.BitList[i];
+		return temp;
+	}
+
+	else if (BitList.GetSize() > rhs.BitList.GetSize())
+	{
+		BitMask temp(BitsNumber);
+		for (size_t i = 0; i < rhs.BitList.GetSize(); i++)
+			temp.BitList[i] = BitList[i] | rhs.BitList[i];
+		for (size_t i = rhs.BitList.GetSize(); i < BitList.GetSize(); i++)
+			temp.BitList[i] = BitList[i];
+		return temp;
+	}
+
+	else if (BitList.GetSize() < rhs.BitList.GetSize())
+	{
+		BitMask temp(rhs.BitsNumber);
+		for (size_t i = 0; i < BitList.GetSize(); i++)
+			temp.BitList[i] = BitList[i] | rhs.BitList[i];
+		for (size_t i = BitList.GetSize(); i < rhs.BitList.GetSize(); i++)
+			temp.BitList[i] = rhs.BitList[i];
+		return temp;
+	}
+	return BitMask(0);
+}
+
+BitMask BitMask::operator^(const BitMask& rhs) const
+{
+	if (BitList.GetSize() == rhs.BitList.GetSize())
```

Comment:
All three operators (&, |, ^) can be written much simpler. Core algorithm logic looks like this:
1. newBitsNumber = max(BitsNumber, rhs.BitsNumber)
2. Perform logic operation on values
3. Trim returned BitMask to size == position of the leftmost bit set to 1.

We can discuss this further via pm.